e2e/testcases/sync_ordering_test.go

```diff
@@ -438,44 +442,145 @@ func TestDependencyWithReconciliation(t *testing.T) {
+			[]nomostest.Predicate{
+				pod2SyncPredicate,
+				podCachePredicate(pod2),
+				nomostest.StatusEquals(nt, kstatus.CurrentStatus),
+			},
+			nomostest.WatchTimeout(nt.DefaultWaitTimeout*2))
+	})
+	// Watch in the background
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+		errCh <- tg.Wait()
+	}()
+
+	// Wait for both watches to be synchronized.
+	// This means it knows the ResourceVersion from before the following changes.
+	<-pod1SyncCh
+	<-pod2SyncCh
+
 	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add pod1 and pod2 (pod2 depends on pod1)")
 	nt.WaitForRepoSyncs()
 
-	var pod1, pod2 *corev1.Pod
-	_, err := nomostest.Retry(30*time.Second, func() error {
-		pod1 = &corev1.Pod{}
-		pod2 = &corev1.Pod{}
-		var errs error
-		errs = multierr.Append(errs, nt.Validate(pod1Name, namespaceName, pod1, isPodReady))
-		errs = multierr.Append(errs, nt.Validate(pod2Name, namespaceName, pod2, isPodReady))
-		return errs
-	})
-	if err != nil {
+	nt.T.Logf("Wait for both pods to become ready (foreground)")
+	if err := <-errCh; err != nil {
 		nt.T.Fatal(err)
 	}
 
-	nt.T.Logf("Verify that pod2 is created after pod1 is ready")
-	readyTime := getPodReadyTimestamp(pod1)
-	if pod2.CreationTimestamp.Before(readyTime) {
-		nt.T.Fatalf("an object (%s) should be created after its dependency (%s) is ready", core.GKNN(pod2), core.GKNN(pod1))
+	// Apply order: pod1 -> pod2 (pod2 depends on pod1)
+	nt.T.Logf("Verify that pod2 was created after pod1 was ready")
+	if pod2.CreationTimestamp.Before(getPodReadyTimestamp(pod1)) {
+		nt.T.Fatalf("an object (%s) should be created after its dependency (%s) is ready",
+			core.GKNN(pod2), core.GKNN(pod1))
 	}
 
 	nt.T.Logf("Remove Pod1 and Pod2")
 	nt.RootRepos[configsync.RootSyncName].Remove("acme/pod1.yaml")
 	nt.RootRepos[configsync.RootSyncName].Remove("acme/pod2.yaml")
+
+	pod1 = &corev1.Pod{}
+	pod2 = &corev1.Pod{}
+	pod1SyncPredicate, pod1SyncCh = nomostest.WatchSyncPredicate()
+	pod2SyncPredicate, pod2SyncCh = nomostest.WatchSyncPredicate()
+
+	// Delete order: pod2 -> pod1 (pod2 depends on pod1)
+	// Unfortunately, there's no "not found timestamp" to compare with the
+	// deletion timestamp, so we're using `metadata.managedFields[*].time` as
+	// a proxy for the "last update timestamp".
+	// Note: We can't use client-side timestamps here, because the events are
+	// not always received in the same order they happened, between two watches.
+	var pod1DeletionTimestamp time.Time   // when pod1 is first MODIFIED to have a DeletionTimestamp
+	var pod2LastUpdateTimestamp time.Time // when pod2 is last MODIFIED before being DELETED
+	pod1DeletionPredicate := func(obj client.Object) error {
+		if obj != nil && pod1DeletionTimestamp.IsZero() && !obj.GetDeletionTimestamp().IsZero() {
+			pod1DeletionTimestamp = obj.GetDeletionTimestamp().Time
+		}
+		return nil
+	}
+	pod2LastUpdatedPredicate := func(obj client.Object) error {
+		if obj != nil && pod2LastUpdateTimestamp.IsZero() {
```

Comment:
Why do we only enter this if `pod2LastUpdateTimestamp.IsZero()` is true? I assume this means we only update this the first time we get a nonzero `LastUpdateTimestamp`, but my understanding is that we want to update this until we see the final `LastUpdateTimestamp`

Reply:
Fixed